test/gtest/ucp/ucp_test.cc

```diff
@@ -583,13 +577,57 @@ void ucp_test_base::entity::fence(int worker_index) const {
     ASSERT_UCS_OK(status);
 }
 
-void* ucp_test_base::entity::disconnect_nb(int worker_index, int ep_index,
-                                           enum ucp_ep_close_mode mode) const {
+void *ucp_test_base::entity::disconnect_nb(int worker_index, int ep_index,
+                                           enum ucp_ep_close_mode mode) {
     ucp_ep_h ep = revoke_ep(worker_index, ep_index);
     if (ep == NULL) {
         return NULL;
     }
-    return ucp_ep_close_nb(ep, mode);
+
+    void *req = ucp_ep_close_nb(ep, mode);
+    if (UCS_PTR_IS_PTR(req)) {
+        m_close_ep_reqs.push_back(req);
+        return req;
+    }
+
+    ASSERT_UCS_OK(UCS_PTR_STATUS(req));
+    return NULL;
+}
+
+bool ucp_test_base::entity::is_ep_closed(void *close_req) const {
+    return (close_req == NULL) ||
+           (ucp_request_check_status(close_req) != UCS_INPROGRESS);
+}
+
+void ucp_test_base::entity::close_ep_req_free(void *close_req) {
+    ASSERT_TRUE(is_ep_closed(close_req)) << "free not completed EP close request";
+    ASSERT_TRUE((close_req == NULL) ||
+                (ucp_request_check_status(close_req) == UCS_OK)) << "ucp_ep_close_nb failed";
+
+    if (close_req == NULL) {
+        return;
+    }
+
+    m_close_ep_reqs.erase(std::find(m_close_ep_reqs.begin(),
+                                    m_close_ep_reqs.end(), close_req));
+    ucp_request_free(close_req);
+}
+
+void ucp_test_base::entity::close_all_eps(const ucp_test &test, int worker_idx,
+                                          enum ucp_ep_close_mode mode) {
+    for (int j = 0; j < get_num_eps(worker_idx); j++) {
+        disconnect_nb(worker_idx, j, mode);
+    }
+
+    ucs_time_t deadline = ucs::get_deadline();
+    while (!m_close_ep_reqs.empty() && (ucs_get_time() < deadline)) {
```

Comment:
shall we fail the test if deadline exceeded?

Reply:
fixed